osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs

```diff
@@ -29,17 +29,35 @@ public partial class SpinnerRotationTracker : CircularContainer
 
         private bool rotationTransferred;
 
-        [Resolved(canBeNull: true)]
+        [Resolved]
         private IGameplayClock? gameplayClock { get; set; }
 
+        private Stack<(double time, float rotation)>? rotationHistory;
+
         public SpinnerRotationTracker(DrawableSpinner drawableSpinner)
         {
             this.drawableSpinner = drawableSpinner;
-            drawableSpinner.HitObjectApplied += resetState;
 
             RelativeSizeAxes = Axes.Both;
         }
 
+        [BackgroundDependencyLoader]
+        private void load(IFrameStableClock? frameStableClock)
+        {
+            if (frameStableClock == null)
```

Comment:
How does this guard work? I'm asking, because I just checked this branch out, and it seems it... doesn't? I started the game up, picked a random map with a spinner in the beginning, right-clicked it, selected "Edit", breakpointed here, and...

![1685386514](https://github.com/ppy/osu/assets/20418176/d30a2d38-caae-4a8f-a722-2b52b840babc)

I basically can't see this change having any effect at all. I even tried applying

```diff
diff --git a/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs b/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
index 8352a28f00..414f6b1df2 100644
--- a/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
+++ b/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
@@ -44,7 +44,7 @@ public SpinnerRotationTracker(DrawableSpinner drawableSpinner)
         [BackgroundDependencyLoader]
         private void load(IFrameStableClock? frameStableClock)
         {
-            if (frameStableClock == null)
+            if (frameStableClock == null || (frameStableClock as FrameStabilityContainer)?.FrameStablePlayback == false)
             {
                 // This class already guarantees correct rewind behaviour via the rate-adjusted calculations, but
                 // this only holds as long as spinners are within a frame-stable playback environment.
diff --git a/osu.Game/Rulesets/UI/FrameStabilityContainer.cs b/osu.Game/Rulesets/UI/FrameStabilityContainer.cs
index 4bb145973d..11f70719f2 100644
--- a/osu.Game/Rulesets/UI/FrameStabilityContainer.cs
+++ b/osu.Game/Rulesets/UI/FrameStabilityContainer.cs
@@ -32,7 +32,7 @@ public sealed partial class FrameStabilityContainer : Container, IHasReplayHandl
         /// <summary>
         /// Whether to enable frame-stable playback.
         /// </summary>
-        internal bool FrameStablePlayback { get; set; } = true;
+        public bool FrameStablePlayback { get; set; } = true;
 
         protected override bool RequiresChildrenUpdate => base.RequiresChildrenUpdate && state != PlaybackState.NotValid;
 
```

and that *also* doesn't work, because `DrawableEditorRulesetWrapper` is not guaranteed to set `FrameStablePlayback` to false before the objects have loaded in.

Reply:
Aha, this is a touch unfortunate.

I used to have this code inside `Update` and was looking for a less ugly solution. Will try and find another way.